### DISCUSSION TITLE
publicRuntimeConfig seems to always be empty

### DISCUSSION THREAD

**Discussion by creeefs** [vercel/next.js#27896] [Help]
### What version of Next.js are you using?

11.0.1

### What version of Node.js are you using?

16.6.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Docker

### Describe the Bug

For some reason, the `publicRuntimeConfig` seems to always be empty.

#### `next.config.js`
```js
module.exports = {
  reactStrictMode: true,
  publicRuntimeConfig: {
    aStaticValue: "hello",
    aRuntimeEnvVar: process.env.NEXT_PUBLIC_HELLO,
  },
};
```

#### Disable static optimization for any `/pages/page.tsx`
```js
const { publicRuntimeConfig } = getConfig();

// always prints an empty object
console.log(publicRuntimeConfig);

// Disable static optimization and use server-side rendering bc of runtime dependency on env vars
export async function getServerSideProps() {
  return { props: {} };
}
```

### Expected Behavior

`publicRuntimeConfig` should return static values as well as environment variables

### To Reproduce

See description for reproduction steps

**Comment by deema089786** [comment]
Reproduced in Next 12.0.3 as well

**Comment by fdev** [comment]
Did you include the `next.config.js` file in your Docker image? It won't work without it.

**Comment by nx-jan-fischer** [comment]
On this page https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration you can read: **A page that relies on publicRuntimeConfig must use getInitialProps to opt-out of Automatic Static Optimization. Runtime configuration won't be available to any page (or component in a page) without getInitialProps.**

So we have to use getInitialProps to use runtime environment variables? To me env vars only make sense if I can change them on the serverside without a rebuild of the application. I would like to use e.g. the same docker image on different environments.

**Reply by Ex4l** [reply]
Hello Jan, did you figure out a way to work with this ? 

We have exactly the same behaviour with one build and have one docker image for x different environnements that we used to feed with different runtime configs but it's totally not working in Next 12.

**Reply by nx-jan-fischer** [reply]
Currently we build the docker image on each environment. But this could be a solution: https://github.com/andrewmclagan/react-env

**Reply by Ex4l** [reply]
Thanks for the answer. Unfortunately we won't be able to do one build per env but i'll take a look at react-env.

**Reply by nx-jan-fischer** [reply]
But remember if you do static site generation at build time the static content from the build environment is embedded into the docker image. With ISR this should be no problem.

**Reply by Ex4l** [reply]
We don't use ISR, we just want to be able to use PublicRuntimeConfig the way we used to do it (and that was working perfectly) in Next 11 😓

**Comment by nx-jan-fischer** [comment]
Another thing that I just discovered and that costed me one day. If you use publicRuntimeConfig and have an environment variable during build, it is hardcoded into the build output (.next/required-server-files.json).

So in your case you don't have an environment variable during build, that's why it's undefined and cannot be set during runtime because you are not using `getInitialProps`.

@timneutkens Can you verify this? Is there another way we can have runtime environment variables? What about the proposed way from Remix to pass the vars from server to the window object? https://remix.run/docs/en/v1/guides/envvars

**Reply by mltsy** [reply]
We have multiple variables in publicRuntimeConfig (feature flags mostly) that are not being frozen at build time (they to respond to ENV changes in staging and production environments).  Variables prepended with `NEXT_PUBLIC_` do have this drawback of being frozen at build time, at least on the client side, so I wonder if that's the issue you were running into?
